File: chr.pos.R
subject <- 
  c("chr10:213,054,000-213,055,000",
    "chrM:111,000-222,000",
    "foo bar",
    "chr1:110-111 chr2:220-222")

## The .* pattern means to match 0 or more of any characters. If there
## are several matches, it will match as much as possible. This does
## not give the desired result for the last subject.
pattern.greedy.stars <- paste0(
  "(?<chrom>chr.*)",
  ":",
  "(?<chromStart>.*)",
  "-",
  "(?<chromEnd>.*)")
str_match_perl(subject, pattern.greedy.stars)

## Using a character class [] assumes you know all possible values
## that you want to match. May not match all possible things you want
## to match (e.g. chrUn_ does not match the chrom group below).
pattern.greedy.class <- paste0(
  "(?<chrom>chr[0-9XYM]+)",
  ":",
  "(?<chromStart>[0-9,]+)",
  "-",
  "(?<chromEnd>[0-9,]+)")
str_match_perl(subject, pattern.greedy.class)

## You can specify the pattern using a negated character class
## (beginning with ^) but that is repetitive. For example you have to
## specify the : separator twice (once in the negated character class,
## and once after the group).
pattern.greedy.negated <- paste0(
  "(?<chrom>chr[^:]+)",
  ":",
  "(?<chromStart>[^-]+)",
  "-",
  "(?<chromEnd>[^ ]*)")
str_match_perl(subject, pattern.greedy.negated)

## I prefer using the .*? non-greedy match. Like .* it matches 0 or
## more of any characters, but it prefers the smallest possible match.
pattern.not.greedy <- paste0(
  "(?<chrom>chr.*?)",
  ":",
  "(?<chromStart>.*?)",
  "-",
  "(?<chromEnd>[0-9,]*)")
str_match_perl(subject, pattern.not.greedy)

## Specify a list of conversion functions as the third argument if you
## want a data.frame with non-character columns.
keep.digits <- function(x)as.integer(gsub("[^0-9]", "", x))
conversion.list <- list(chromStart=keep.digits, chromEnd=keep.digits)
str_match_perl(subject, pattern.not.greedy, conversion.list)

## Note that str_match_perl above returned only the first match in
## every subject string. If you want EVERY match, use
## str_match_all_perl instead.
str_match_all_perl(subject, pattern.not.greedy)
str_match_all_perl(subject, pattern.not.greedy, conversion.list)

File: chr_pos.py
import pandas as pd
import re

# Subjects are specified as a list or a pandas Series.
subject_list = [
    "chr10:213,054,000-213,055,000",
    "chrM:111,000-222,000",
    "foo bar",
    None,
    "chr1:110-111 chr2:220-222"
    ]
subject_series = pd.Series(subject_list)

# too greedy, wrong answer on last subject.
pattern_greedy_stars = r"""
(?P<chrom>chr.*)
:
(?P<chromStart>.*)
-
(?P<chromEnd>.*)
"""
subject_series.str.extract(pattern_greedy_stars, re.VERBOSE)

# greedy character class, wrong answer on chrM subject.
pattern_greedy_class = r"""
(?P<chrom>chr[0-9]+)
:
(?P<chromStart>[0-9,]+)
-
(?P<chromEnd>[0-9,]+)
"""
subject_series.str.extract(pattern_greedy_class, re.VERBOSE)

# negated character class, right answer but repetitive.
pattern_greedy_negated = r"""
(?P<chrom>chr[^:]+)
:
(?P<chromStart>[^-]+)
-
(?P<chromEnd>[^ ]*)
"""
subject_series.str.extract(pattern_greedy_negated, re.VERBOSE)

# non-greedy, right answer.
pattern_not_greedy = r"""
(?P<chrom>chr.*?)
:
(?P<chromStart>.*?)
-
(?P<chromEnd>[0-9,]*)
"""
match_df = subject_series.str.extract(pattern_not_greedy, re.VERBOSE)
print match_df

# pandas does not support NaN in .astype(int) so we must use float.
# http://pandas.pydata.org/pandas-docs/stable/gotchas.html#support-for-integer-na
def tofloat(series):
    return series.str.replace(r"[^0-9]", "").astype(float)

match_df["chromStart_float"] = tofloat(match_df["chromStart"])
match_df["chromEnd_float"] = tofloat(match_df["chromStart"])
print match_df

# str.extractall returns a DataFrame with one row for each match and
# one column for each capture group.
extractall_df = subject_series.str.extractall(pattern_not_greedy, re.VERBOSE)
print extractall_df

